kratos/utilities/nodal_sensitivity_builder.h

```diff
@@ -0,0 +1,288 @@
+
+private:
+    ///@name Member Variables
+    ///@{
+
+    ModelPart& mrModelPart;
+    AdjointResponseFunction::Pointer mpResponseFunction;
+
+    ///@}
+    ///@name Private Operations
+    ///@{
+
+    template <class TDataType>
+    void BuildNodalSolutionStepElementContributions(Variable<TDataType> const& rVariable,
+                                                    double ScalingFactor)
+    {
+        KRATOS_TRY;
+
+        auto& r_elements = mrModelPart.Elements();
+        const auto& r_process_info = mrModelPart.GetProcessInfo();
+
+        const int num_threads = OpenMPUtils::GetNumThreads();
+        std::vector<Vector> local_sensitivity(num_threads);
+        std::vector<Vector> partial_sensitivity(num_threads);
+        std::vector<Vector> adjoint_vector(num_threads);
+        std::vector<Matrix> sensitivity_matrix(num_threads);
+
+#pragma omp parallel
+        {
+            ModelPart::ElementIterator elements_begin;
+            ModelPart::ElementIterator elements_end;
+            OpenMPUtils::PartitionedIterators(r_elements, elements_begin, elements_end);
+            int k = OpenMPUtils::ThisThread();
+
+            for (auto it = elements_begin; it != elements_end; ++it)
+            {
+                Element::GeometryType& r_geom = it->GetGeometry();
+
+                if (HasActiveNodes(r_geom) == false)
+                    continue;
+
+                it->CalculateSensitivityMatrix(rVariable, sensitivity_matrix[k], r_process_info);
+
+                mpResponseFunction->CalculatePartialSensitivity(
+                    *it, rVariable, sensitivity_matrix[k],
+                    partial_sensitivity[k], r_process_info);
+
+                it->GetValuesVector(adjoint_vector[k]);
+
+                if (local_sensitivity[k].size() != sensitivity_matrix[k].size1())
+                    local_sensitivity[k].resize(sensitivity_matrix[k].size1(), false);
+
+                noalias(local_sensitivity[k]) =
+                    ScalingFactor * (prod(sensitivity_matrix[k], adjoint_vector[k]) +
+                                     partial_sensitivity[k]);
+
+                AssembleNodalSolutionStepSensitivityContribution(
+                    rVariable, local_sensitivity[k], r_geom);
+            }
+        }
+
+        KRATOS_CATCH("");
+    }
+
+    template <class TDataType>
+    void BuildNodalSolutionStepConditionContributions(Variable<TDataType> const& rVariable,
+                                                      double ScalingFactor)
+    {
+        KRATOS_TRY;
+
+        auto& r_conditions = mrModelPart.Conditions();
+        const auto& r_process_info = mrModelPart.GetProcessInfo();
+
+        const int num_threads = OpenMPUtils::GetNumThreads();
+        std::vector<Vector> local_sensitivity(num_threads);
+        std::vector<Vector> partial_sensitivity(num_threads);
+        std::vector<Vector> adjoint_vector(num_threads);
+        std::vector<Matrix> sensitivity_matrix(num_threads);
+
+#pragma omp parallel
```

Comment:
I think the "official" way to do omp element/condition loops now is to just do for int i = 0 to number_of_elements and operate with Elements.begin()+i
@RiccardoRossi

Reply:
yes it is as you say